twilight-gateway/src/stream.rs

```diff
@@ -1,12 +1,36 @@
-//! Streaming utilities for initializing clusters of shards.
+//! Streaming utilities for initializing groups of shards.
+//!
+//! There are two groups of functionality to the stream module: initializers and
+//! selectors. The initializers are functions like [`start_recommended`], which
+//! initializes a group of shards based on the recommended number of shards
+//! Discord recommends is started for the application. Once these shards are
+//! initialized, the events or websocket messages of all of the shards can be
+//! collected into an efficient stream via [`ShardEventStream`] and
+//! [`ShardMessageStream`].
 
-use crate::{error::ShardInitializeError, Config, Shard, ShardId};
-use futures_util::stream::{FuturesUnordered, Stream};
+use crate::{
+    error::{ReceiveMessageError, ShardInitializeError},
+    message::Message,
+    tls::TlsContainer,
+    Config, Shard, ShardId,
+};
+use futures_util::stream::{FuturesUnordered, Stream, StreamExt};
 use std::{
+    cell::RefCell,
     error::Error,
     fmt::{Display, Formatter, Result as FmtResult},
+    future::Future,
+    ops::{Deref, DerefMut},
+    pin::Pin,
+    rc::Rc,
+    task::{Context, Poll},
 };
 use twilight_http::Client;
+use twilight_model::gateway::event::Event;
+
+/// Generic list of unordered futures producing an item for each shard.
+type FutureList<'a, Item> =
+    FuturesUnordered<Pin<Box<dyn Future<Output = NextItemOutput<'a, Item>> + 'a>>>;
```

Comment:
Just asking a bit here, but can this be done without erasing the type so static dispatch can be used instead?

Reply:
I might need you to elaborate on the request, but with existential types being unstable still we would end up needing dynamic dispatch somewhere anyway due to underlying vampire types (next_event is defined as an `async fn`).